### ISSUE TITLE
Embedding visualizations for hi-res images training (feature request)

### ISSUE THREAD

**Issue by argavish** [tensorflow/tensorflow#8887]
Hi Tensorflow people, thank you all.
I wonder if I could train my image classification model on my normal sized images (1080px X 1920px) but still use Tensorboard's cool embedding visualizations. I prefer to keep the images with the highest resolution possible since it proved to be important for the classification.

In the tutorial (https://www.tensorflow.org/get_started/embedding_viz) the developers specify tensorflow currently supports sprites up to 8192px X 8192px, meaning you can either use a lot of low-res images (fine for MNIST 28X28 and CIFAR-10 32X32) or a few high-res images (a too small training set size). So, I was wondering if there could be a way around it.

**What if, we could use full-res photos for training, but downsize them for the thumbnails needed to make up the sprite?**
That way we could still get a sense of which picture is which in the visualization, but let the model train on higher quality data.

1. Can I do it myself, by creating a low-res (down-sized) copy of my entire database beforehand, and use it to create the sprite etc., will the embedding event still correlate to the same source image (and the right label in the metadata)?
2. Instead, should I put inside the training code itself, a small procedure for resizing of the full-res image after each bottleneck calculation, and then store that thumbnail in a separate folder - making sure the embedding log in the metadata actually corresponds to the right thumbnail?

***************************************
Environment info:

Ubuntu 16.04.02 (64 bit)
tensorflow 0.12.1 CPU only (64 bit). I have a NVIDIA GTX 1050ti waiting to be used if crucial for this task.

I'm am a kind of a coding noob so forgive my inaccuracies and ignorance. I'm relatively new (6 months) to tensorflow and CNNs in general. I've been transfer-training inception V3 on classification of large (1080px X 1920px) images, divided to 10 labels (folders). 
The reason I'm asking and not just diving deep into it is that I'll have to spend a lot of time to resize my images and create the perfect sprite image and metadata file, but won't have the confidence that the trained data corresponds to the sprite image. So I want to see if it's even possible to begin with - **to get full certainty of visualized thumbnail corresponding to actual full-res image used for training.**

Thank you for this great platform!! Tensorboard is a very powerful tool and I'm very excited to unlock the embedding visualizations' potential. @dandelionmane

**Comment by aselle** [comment]
@dandelionmane, could you provide direction on this feature and if it is necessary and how to go about it?

**Comment by argavish** [comment]
Thanks @aselle !

Here's some bits I found that may help (I hope I'm not totally off here):

1. This sprite builder [here](https://github.com/oduerr/dl_tutorial/blob/master/tensorflow/debugging/embedding.ipynb), line [16]. I still need to figure out how to build it straight from the source images (after resizing them) and not a from a tensor already containing the images as used there.

2. Similarly, generating the metadata file, like [here](https://github.com/normanheckscher/mnist-tensorboard-embeddings/blob/master/mnist_t-sne.py), line [69]

3. To refine the question - should the sprite and metadata files be built continuously with every step in the training, or can they be prepared in advance?

Thanks again!

**Comment by PhilJd** [comment]
The sprite sheet for the embedding visualization is independent of your training data. The sprite sheet won't be used as input for the training process. 

To ensure that a sprite of the sprite sheet corresponds to an embedding, you need to make sure that you compute and add the embeddings to tensorboard in the same order that you have in the sprite sheet.
Most probably you'll only add embeddings to tensorboard every x iterations. So you could shuffle your images during training and when you reach iteration x, feed them in order. 

```
projector_config = projector.ProjectorConfig()
emb_conf = self.projector_config.embeddings.add()
emb_conf.metadata_path = "tsv_filename"
emb_conf.sprite.image_path = "spritesheet_name"
train_images = [img1, img2 ...]  # same order as in your spritesheet


inference_op = ... # the inference op of your graph, taking images an input, outputs the embeddings
loss_op = loss(inference_op)  # compute the loss based on your embeddings
train_op = tf.Optimizer().minimize(loss_op)  # the traninig op of your graph
sess = tf.Session()
for step in range(1000):
    sess.run(train_op)
    if step % 100 == 0:
        for img_batch in train_images:
            embeddings += sess.run(inference_op, feed={placeholder: img_batch})
       ... routine to add and save the embeddings with tensorboard
            
```

Regarding your third question: The only situation where creating the sprites during training might be useful is if you alter the images during training. However if you only apply small changes, e.g. add some color noise, the resulting embedding won't be too different and you'd only get a lot of images on the same spot.

**Comment by argavish** [comment]
Thanks for the clarification and code @PhilJd ! 
So, hopefully the low res images in the sprite image will be classified similarly to their full-res sources.
I'll give it a go! Thanks again

**Comment by dandelionmane** [comment]
Closing this issue, since as PhilJd correctly points out, the sprites are independent from the training data.
